src/utils/fee_calculator.ts

```diff
@@ -69,31 +101,112 @@ export function calculateFees(opts: {
+            fees.zeroExFee.feeRecipient &&
+            fees.zeroExFee.type === 'integrator_share'
+        ) {
+            zeroExFeeAdjustment = fees.zeroExFee.feeAmount;
+        }
+
+        feeRecipientToOnChainTransfer.set(fees.integratorFee.feeRecipient, {
+            feeRecipient: fees.integratorFee.feeRecipient,
+            feeToken: fees.integratorFee.feeToken,
+            feeAmount: currentIntegratorFeeAmount.plus(fees.integratorFee.feeAmount.minus(zeroExFeeAdjustment)),
+            affilateFeeType: AffiliateFeeType.PercentageFee,
+        });
     }
     // 0x fee
-    if (fees.zeroExFee && fees.zeroExFee.billingType === 'on-chain' && fees.zeroExFee.feeRecipient) {
-        // If the fee kind is integrator_share, the 0x amount has already been included in integrator amount
-        if (fees.zeroExFee.type !== 'integrator_share') {
-            totalFeeAmount = totalFeeAmount.plus(fees.zeroExFee.feeAmount);
-        }
+    if (
+        fees.zeroExFee &&
+        fees.zeroExFee.billingType === 'on-chain' &&
+        fees.zeroExFee.feeRecipient &&
+        fees.zeroExFee.feeAmount.gt(ZERO)
+    ) {
+        const currentZeroExFeeAmount =
+            feeRecipientToOnChainTransfer.get(fees.zeroExFee.feeRecipient)?.feeAmount ?? ZERO;
+        feeRecipientToOnChainTransfer.set(fees.zeroExFee.feeRecipient, {
+            feeRecipient: fees.zeroExFee.feeRecipient,
+            feeToken: fees.zeroExFee.feeToken,
+            feeAmount: currentZeroExFeeAmount.plus(fees.zeroExFee.feeAmount),
+            affilateFeeType: AffiliateFeeType.PercentageFee,
+        });
     }
     // Gas fee
-    if (fees.gasFee && fees.gasFee.billingType === 'on-chain' && fees.gasFee.feeRecipient) {
-        totalFeeAmount = totalFeeAmount.plus(fees.gasFee.feeAmount);
+    if (
+        fees.gasFee &&
+        fees.gasFee.billingType === 'on-chain' &&
+        fees.gasFee.feeRecipient &&
+        fees.gasFee.feeAmount.gt(ZERO)
+    ) {
+        const currentGasFeeAmount = feeRecipientToOnChainTransfer.get(fees.gasFee.feeRecipient)?.feeAmount ?? ZERO;
+        feeRecipientToOnChainTransfer.set(fees.gasFee.feeRecipient, {
+            feeRecipient: fees.gasFee.feeRecipient,
+            feeToken: fees.gasFee.feeToken,
+            feeAmount: currentGasFeeAmount.plus(fees.gasFee.feeAmount),
+            affilateFeeType: AffiliateFeeType.GaslessFee,
+        });
+    }
+
+    let onChainTransfersGas = ZERO;
+    switch (onChainFeeTransferType) {
+        case 'transferFrom':
+            onChainTransfersGas = TRANSFER_FROM_GAS.times(feeRecipientToOnChainTransfer.size);
+            break;
+        case 'affiliateFee':
+            onChainTransfersGas = AFFILIATE_FEE_TRANSFORMER_GAS.plus(TRANSFER_FROM_GAS).times(
+                feeRecipientToOnChainTransfer.size,
+            );
+            break;
+        default:
+            ((_: never) => {
+                throw new Error('unreachable');
+            })(onChainFeeTransferType);
```

Comment:
I feel like the `onChainFeeTransferType` abstraction is a lot of complexity to add the `AFFILIATE_FEE_TRANSFORMER_GAS` overhead. I would recommend we just ignore this overhead and just underquote gas (especially since it will be short lived)

Reply:
Agree. Another use for `onChainFeeTransferType` (other than determining we're going to use affiliate fee transformer or use `transferFrom` directly to transfer fee) is to decide if we'd like to use `TRANSFER_GAS` (used by affiliate fee transfer) or `TRANSFER_FROM_GAS` (used by `transferFrom`). Here I made a typo. It should be:
```        case 'transferFrom':
            onChainTransfersGas = TRANSFER_FROM_GAS.times(...);
            break;
        case 'affiliateFee':
            onChainTransfersGas = AFFILIATE_FEE_TRANSFORMER_GAS.plus(TRANSFER_GAS).times(
               ...
            );
```

Reply:
Maybe the caller should just pass the gas cost for an on-chain transfer in as a parameter? That would simply the logic a lot